op-chain-ops/cmd/celo-migrate/beacon.go

```diff
@@ -0,0 +1,254 @@
+			continue
+		}
+		if err != nil {
+			return common.Hash{}, fmt.Errorf("error fetching beacon block at slot %d: %w", mostRecentFinalizedSlot, err)
+		}
+		if !beaconBlock.Finalized {
+			return common.Hash{}, fmt.Errorf("expecting beacon block at slot %d to be finalized", mostRecentFinalizedSlot)
+		}
+		break // We found a good block.
+	}
+	if beaconBlock == nil {
+		return common.Hash{}, fmt.Errorf("failed to find finalized block searching up to 10 slots back from the most recent finalized slot (%d) at the L2 fork time (%d)", mostRecentFinalizedSlot, l2StartTimeSeconds)
+	}
+	return common.HexToHash(beaconBlock.Data.Message.Body.ExecutionPayload.BlockHash), nil
+}
+
+// Gets the epoch from the beaconcha.in api
+func (c *beaconClient) Epoch(ctx context.Context, num uint64) (epoch *Epoch, err error) {
+	headers := http.Header{}
+	headers.Add("Accept", "application/json")
+	resp, err := c.cl.Get(fmt.Sprintf("%s/v1/epoch/%d", c.beaconchainURL, num))
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch epoch: %w", err)
+	}
+	defer func() {
+		err = errors.Join(err, resp.Body.Close())
+	}()
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("failed to fetch epoch, http status %d", resp.StatusCode)
+	}
+	epoch = &Epoch{}
+	if err := json.NewDecoder(resp.Body).Decode(epoch); err != nil {
+		return nil, fmt.Errorf("failed to decode epoch: %w", err)
+	}
+	return epoch, nil
+}
+
+// Gets the beacon block from the beacon rpc api
+func (c *beaconClient) BeaconBlock(ctx context.Context, slot uint64) (block *BeaconBlock, err error) {
+	headers := http.Header{}
+	headers.Add("Accept", "application/json")
+	resp, err := c.cl.Get(fmt.Sprintf("%s/eth/v2/beacon/blocks/%d", c.beaconRPC, slot))
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch becon block: %w", err)
+	}
+	defer func() {
+		err = errors.Join(err, resp.Body.Close())
+	}()
+	if resp.StatusCode != http.StatusOK {
+		if resp.StatusCode == http.StatusNotFound {
+			return nil, fmt.Errorf("failed to fetch beacon block at slot %d: %w", slot, ethereum.NotFound)
+		}
+		return nil, fmt.Errorf("failed to fetch beacon block at slot %d, http status %d", slot, resp.StatusCode)
+	}
+	if err := json.NewDecoder(resp.Body).Decode(&block); err != nil {
+		return nil, fmt.Errorf("failed to decode beacon block: %w", err)
+	}
+	return block, nil
+}
+
+// Returns the start time of an epoch
+func EpochStartTime(epoch uint64) uint64 {
+	return beaconChainGenesisTimeSeconds + epoch*beaconSlotsPerEpoch*beaconChainSlotDurationSeconds
+}
+
+// Returns the number of the epoch starting at or before the given time.
+func EpochAtOrBefore(unixTime uint64) uint64 {
+	return SlotAtOrBefore(unixTime) / beaconSlotsPerEpoch
+}
+
+func SlotAtOrBefore(unixTime uint64) uint64 {
+	// Get the slot at or before the given time.
+	// Slot = (start - genesis) / slotDuration
+	return (unixTime - beaconChainGenesisTimeSeconds) / beaconChainSlotDurationSeconds
+}
+
+type BeaconBlock struct {
```

Comment:
Is there some lib for that kind of stuff?

Reply:
I found these two options:
* https://github.com/gobitfly/eth2-beaconchain-explorer
* https://github.com/prysmaticlabs/prysm

But they both just provide clients for the beacon-rpc, so I'd still need to write a client for the beaconcha.in requests, it didn't seem worth it to import all that stuff for one method.

Reply:
Sounds interesting, I will have a look, where is the swagger file?

Reply:
https://ethereum.github.io/beacon-APIs/releases/v3.0.0/beacon-node-oapi.json

The beacon-api specs is basically just a render of that